pkg/controller/networkpolicy/status_controller.go

```diff
@@ -0,0 +1,343 @@
+		return nil
+	}
+	statuses := make([]*controlplane.NetworkPolicyNodeStatus, 0, len(c.statuses[key]))
+	for _, status := range statusPerNode {
+		statuses = append(statuses, status)
+	}
+	return statuses
+}
+
+func (c *StatusController) clearStatuses(key string) {
+	c.statusesLock.Lock()
+	defer c.statusesLock.Unlock()
+	delete(c.statuses, key)
+}
+
+func (c *StatusController) deleteNodeStatus(key string, nodeName string) {
+	c.statusesLock.Lock()
+	defer c.statusesLock.Unlock()
+	statusPerNode, exists := c.statuses[key]
+	if !exists {
+		return
+	}
+	delete(statusPerNode, nodeName)
+}
+
+// Run begins watching and syncing of a StatusController.
+func (c *StatusController) Run(stopCh <-chan struct{}) {
+	defer c.queue.ShutDown()
+
+	klog.Info("Starting NetworkPolicyStatus controller")
+	defer klog.Info("Shutting down NetworkPolicyStatus controller")
+
+	klog.Info("Waiting for caches to sync for NetworkPolicyStatus controller")
+	if !cache.WaitForCacheSync(stopCh, c.cnpListerSynced, c.anpListerSynced) {
+		klog.Error("Unable to sync CNP caches for NetworkPolicy controller")
+		return
+	}
+	klog.Info("Caches are synced for NetworkPolicyStatus controller")
+
+	go wait.NonSlidingUntil(c.watchInternelNetworkPolicy, 5*time.Second, stopCh)
+
+	for i := 0; i < defaultWorkers; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+func (c *StatusController) watchInternelNetworkPolicy() {
+	watcher, err := c.internalNetworkPolicyStore.Watch(context.TODO(), "", labels.Everything(), fields.Everything())
+	if err != nil {
+		klog.Errorf("Failed to start watch for internal NetworkPolicy: %v", err)
+		return
+	}
+	defer watcher.Stop()
+	resultCh := watcher.ResultChan()
+	for {
+		select {
+		case event, ok := <-resultCh:
+			if !ok {
+				return
+			}
+			// Skip handling Bookmark events.
+			if event.Type == watch.Bookmark {
+				continue
+			}
+			np := event.Object.(*controlplane.NetworkPolicy)
+			if np.SourceRef.Type == controlplane.K8sNetworkPolicy {
+				continue
+			}
+			c.queue.Add(np.Name)
+		}
+	}
+}
+
+func (c *StatusController) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem deals with one key off the queue.  It returns false when it's time to quit.
+func (c *StatusController) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	err := c.syncHandler(key.(string))
+	if err == nil {
+		c.queue.Forget(key)
+		return true
+	}
+
+	klog.Errorf("Failed to sync NetworkPolicy status %s: %v", key, err)
+	c.queue.AddRateLimited(key)
+
+	return true
+}
+
+// syncHandler calculates the NetworkPolicy status based on the desired state from the internalNetworkPolicyStore and
+// the actual state from the statuses map, and syncs it with the Kubernetes API.
+func (c *StatusController) syncHandler(key string) error {
+	klog.V(2).Infof("Syncing NetworkPolicy status for %s", key)
+	internalNPObj, found, _ := c.internalNetworkPolicyStore.Get(key)
```

Comment:
Remind me if the object got from the store is immutable or not (no other thread will change it?).

Reply:
immutable, networkpolicy controller will create a new object when it updates one.

Reply:
Ok